src/stream.js

```diff
@@ -0,0 +1,488 @@
+  }
+
+  // Minimum browser compatibility based on current code:
+  //   Chrome                52
+  //   Edge                  79
+  //   Firefox               65
+  //   IE                    NA
+  //   Opera                 39
+  //   Safari                10.1
+  //   Android Webview       52
+  //   Chrome for Android    52
+  //   Firefox for Android   65
+  //   Opera for Android     41
+  //   Safari on iOS         10.3
+  //   Samsung Internet      6.0
+  function platformSpecificEventRead(response) {
+    try {
+      if (util.isNodeEnv()) {
+        response.body.on('data', onData).on('error', onError)
+      } else {
+        // ATENTION: The following code is meant to run on browsers and it's not
+        // covered by current test automation. Manual testing on major browsers
+        // is required after making changes to it.
+        var reader = response.body.getReader()
+        var decoder = new TextDecoder('utf-8')
+        function pump() {
+          return reader
+            .read()
+            .then(function process(msg) {
+              if (!msg.done) {
+                onData(decoder.decode(msg.value))
+                return pump()
+              }
+            })
+            .catch(onError)
+        }
+        pump()
+      }
+    } catch (err) {
+      throw new errors.StreamsNotSupported(
+        'Unexpected error during stream initialization: ' + err
+      )
+    }
+  }
+
+  self._client._http
+    .execute('POST', 'stream', body, self._urlParams, {
+      fetch: self._fetch,
+      signal: self._abort.signal,
+    })
+    .then(function(response) {
+      return handleResponse(response).then(function() {
+        platformSpecificEventRead(response, onData, onError)
+      })
+    })
+    .catch(onError)
+}
+
+/** Closes the stream subscription by aborting its underlying http request. */
+StreamClient.prototype.close = function() {
+  if (this._state !== 'closed') {
+    this._state = 'closed'
+    this._abort.abort()
+  }
+}
+
+/**
+ * Event dispatch interface for stream subscription.
+ *
+ * @constructor
+ * @param {string[]} allowedEvents List of allowed events.
+ * @private
+ */
+function EventDispatcher(allowedEvents) {
+  this._allowedEvents = allowedEvents
+  this._listeners = {}
+}
+
+/** Subscribe to an event
+ *
+ * @param {string} type The event type.
+ * @param {module:stream~Subscription~eventCalllback} callback
+ *   The event's callback.
+ */
+EventDispatcher.prototype.on = function(type, callback) {
+  if (this._allowedEvents.indexOf(type) === -1) {
+    throw new Error('Unkown event type: ' + type)
+  }
+  if (this._listeners[type] === undefined) {
+    this._listeners[type] = []
+  }
+  this._listeners[type].push(callback)
+}
+
+/**
+ * Dispatch the given event to the appropriate listeners.
+ *
+ * @param {Object} event The event.
+ */
+EventDispatcher.prototype.dispatch = function(event) {
+  var listeners = this._listeners[event.event] || []
+  for (var i = 0; i < listeners.length; i++) {
```

Comment:
@erickpintor If `listeners` is an empty array, can we do an early return here and avoid the `for` altogether?